django_sharding_library/manager.py

```diff
@@ -0,0 +1,73 @@
+from django.db import transaction, router, IntegrityError
+from django.db.models.manager import Manager
+from django.db.models.query import QuerySet
+
+
+class ShardQuerySet(QuerySet):
+    """
+    Stores the lookups to pass to the router as a hint
+    """
+    def __init__(self, model=None, *args, **kwargs):
+        super(ShardQuerySet, self).__init__(model=model, *args, **kwargs)
+        self._exact_lookups = {}
+
+    def _clone(self, **kwargs):
+        clone = super(ShardQuerySet, self)._clone(**kwargs)
+        clone._exact_lookups = self._exact_lookups.copy()
+        return clone
+
+    def _filter_or_exclude(self, *args, **kwargs):
+        """
+        Update our lookups when we get a filter or an exclude (we only care about filter, but its a
+        shared function in the ORM)
+        :return:
+        """
+        clone = super(ShardQuerySet, self)._filter_or_exclude(*args, **kwargs)
+        if getattr(clone, '_exact_lookups', None) is None:
+            clone._exact_lookups = {}
+        clone._exact_lookups.update(dict([(k, v) for k, v in kwargs.items() if '__' not in k]))
+        return clone
+
+    @property
+    def db(self):
+        if self._db:
+            return self._db
+
+        self._hints['exact_lookups'] = self._exact_lookups
+        if not self._hints.get('instance') and getattr(self, '_instance', None):
+            self._hints['instance'] = getattr(self, '_instance')
+
+        return super(ShardQuerySet, self).db
+
+    def create(self, **kwargs):
+        """
+        Not sure if this is necessary anymore, it was in 1.4. Grabs the instance before its too late to pass it to the
```

Comment:
It is required. I will remove the comment. Django does a terrible job of keeping track of the instance for some reason unless you grabbed it from the database already (e.g.: create should have an instance, but it doesn't because Django doesn't keep the instance around until after the save() is run.) I am actually looking at putting a PR in to django to change this since it seems arbitrary, sent a message to the mailing list a few weeks ago.